### PR TITLE
Puppi isolation miniAOD electrons & photons v2

### PR DESCRIPTION
**Pull request by ishvetso**
PUPPI isolation for electrons and photons for miniAOD. 
this is a rebase of #16055

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@nhanvtran @violatingcp @gzevi please check/confirm that photons do not (and will not) need a puppiNoLep isolations

**Comment by nhanvtran** [comment]
@slava77, photons do not benefit from using puppiNoLep in the isolation -- they are practically the same as puppi with leptons.  We would need an analog of puppiNoPhotons but @ishvetso has not studied a possible "puppiNoPhotons" improvement yet.  That said, the current implementation already brings something so we prefer to stick with the validated version for now.

**Comment by gzevi** [comment]
@slava77: as far as I know, puppiNoLep isolation for photons has never been tested. I don't know enough to answer the question about future possible usage, but if this is the case I would prefer not to add a new variable that has never been looked at.

**Comment by monttj** [comment]
+1
It was already signed here #16055.

File: DataFormats/PatCandidates/interface/Photon.h [hunk 1]
```diff
@@ -43,6 +43,7 @@ namespace reco {
 // Class definition
 namespace pat {
   class PATPhotonSlimmer;
+  class AssociatedCandProducer;
```
Comment (ishvetso) [inline DataFormats/PatCandidates/interface/Photon.h]: fixed

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/16172 af5583c
- code changes are in line with the PR description and the follow up code review. For miniAOD electrons 6 new isolation floats are added and 3 new floats are added for photons.
- jenkins tests pass and comparisons with baseline show no differences as expected (based on previous commit tests)
- local tests with 1000 ttbar PU35 events show that new variables are filled with somewhat reasonable distributions for isolations. On the technical side: 
    - the event size increased by 55 bytes/evt in slimmedPhotons and slimmedElectrons (with a few bytes of compression noise) or by 0.1%; 
    - egmPhotonPUPPIIsolationForPhotons  module is gone to be replaced by egmElectronPUPPIIsolation, egmElectronPUPPINoLeptonsIsolation, and egmPhotonPUPPIIsolation for a net increase in CPU time of about 3 ms/evt (0.15% of miniAOD processing time).

**Comment by monttj** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**